### PR TITLE
#12537 : Added fmin mathematical function for numpy frontend

### PR DESCRIPTION
**Pull request by AryanSharma9917**

### CODE REVIEW

**Comment by AryanSharma9917** [comment]
Heyy @kurshakuz , how do I solve this lint issue could you please guide me out

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py [hunk 1]
```diff
@@ -57,6 +57,45 @@ def test_numpy_minimum(
         subok=True,
     )
 
+# fmin
+@handle_frontend_test(
+    fn_tree="numpy.fmin",
+    dtype_x1_x2_axis=helpers.dtype_values_axis2(
```
Comment (kurshakuz) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:63]: This seems to be an error right?
```suggestion
    dtype_x1_x2_axis=helpers.dtype_values_axis(
```

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hey! I have tested your implementation locally and it fails for all 4 tests. Additionally, you have a syntax error that was causing one of the issues. Please fix your solution, test it locally and let me know once you are done, so that I can have a second look. Thanks! 🙂 

```
ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py FFFF                                               [100%]

======================================================================== FAILURES =========================================================================
_____________________________________________ test_numpy_fmin[cpu-ivy.functional.backends.numpy-False-False] ______________________________________________

frontend = 'numpy', on_device = 'cpu'

    @handle_frontend_test(
>       fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dtype_x1_x2_axis = (['float16'], [array([-1.], dtype=float16)], 0), frontend = 'numpy'
test_flags = num_positional_args=1. with_out=False. inplace=False. native_arrays=[False]. as_variable=[False]. 
fn_tree = 'ivy.functional.frontends.numpy.fmin', on_device = 'cpu', where = [array([False])]

    @handle_frontend_test(
        fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )
    def test_numpy_fmin(
        dtype_x1_x2_axis,
        frontend,
        test_flags,
        fn_tree,
        on_device,
        where,
    ):
>       input_dtypes, x1, x2, axis = dtype_x1_x2_axis
E       ValueError: not enough values to unpack (expected 4, got 3)
E       Falsifying example: test_numpy_fmin(
E           dtype_x1_x2_axis=(['float16'], [array([-1.], dtype=float16)], 0),
E           where=[array([False])],
E           fn_tree='ivy.functional.frontends.numpy.fmin',
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=1,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E           ),
E           frontend='numpy',
E           on_device='cpu',
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.68.2', b'AXicY2BkAAMoBacBAEcABA==') as a decorator on your test case

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:82: ValueError
______________________________________________ test_numpy_fmin[cpu-ivy.functional.backends.jax-False-False] _______________________________________________

frontend = 'numpy', on_device = 'cpu'

    @handle_frontend_test(
>       fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dtype_x1_x2_axis = (['float16'], [array([-1.], dtype=float16)], 0), frontend = 'numpy'
test_flags = num_positional_args=1. with_out=False. inplace=False. native_arrays=[False]. as_variable=[False]. 
fn_tree = 'ivy.functional.frontends.numpy.fmin', on_device = 'cpu', where = [array([False])]

    @handle_frontend_test(
        fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )
    def test_numpy_fmin(
        dtype_x1_x2_axis,
        frontend,
        test_flags,
        fn_tree,
        on_device,
        where,
    ):
>       input_dtypes, x1, x2, axis = dtype_x1_x2_axis
E       ValueError: not enough values to unpack (expected 4, got 3)
E       Falsifying example: test_numpy_fmin(
E           dtype_x1_x2_axis=(['float16'], [array([-1.], dtype=float16)], 0),
E           where=[array([False])],
E           fn_tree='ivy.functional.frontends.numpy.fmin',
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=1,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E           ),
E           frontend='numpy',
E           on_device='cpu',
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.68.2', b'AXicY2BkAAMoBacBAEcABA==') as a decorator on your test case

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:82: ValueError
___________________________________________ test_numpy_fmin[cpu-ivy.functional.backends.tensorflow-False-False] ___________________________________________

frontend = 'numpy', on_device = 'cpu'

    @handle_frontend_test(
>       fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dtype_x1_x2_axis = (['float16'], [array([-1.], dtype=float16)], 0), frontend = 'numpy'
test_flags = num_positional_args=1. with_out=False. inplace=False. native_arrays=[False]. as_variable=[False]. 
fn_tree = 'ivy.functional.frontends.numpy.fmin', on_device = 'cpu', where = [array([False])]

    @handle_frontend_test(
        fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )
    def test_numpy_fmin(
        dtype_x1_x2_axis,
        frontend,
        test_flags,
        fn_tree,
        on_device,
        where,
    ):
>       input_dtypes, x1, x2, axis = dtype_x1_x2_axis
E       ValueError: not enough values to unpack (expected 4, got 3)
E       Falsifying example: test_numpy_fmin(
E           dtype_x1_x2_axis=(['float16'], [array([-1.], dtype=float16)], 0),
E           where=[array([False])],
E           fn_tree='ivy.functional.frontends.numpy.fmin',
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=1,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E           ),
E           frontend='numpy',
E           on_device='cpu',
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.68.2', b'AXicY2BkAAMoBacBAEcABA==') as a decorator on your test case

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:82: ValueError
_____________________________________________ test_numpy_fmin[cpu-ivy.functional.backends.torch-False-False] ______________________________________________

frontend = 'numpy', on_device = 'cpu'

    @handle_frontend_test(
>       fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dtype_x1_x2_axis = (['float16'], [array([-1.], dtype=float16)], 0), frontend = 'numpy'
test_flags = num_positional_args=1. with_out=False. inplace=False. native_arrays=[False]. as_variable=[False]. 
fn_tree = 'ivy.functional.frontends.numpy.fmin', on_device = 'cpu', where = [array([False])]

    @handle_frontend_test(
        fn_tree="numpy.fmin",
        dtype_x1_x2_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("float"),
            min_num_dims=1,
            valid_axis=True,
            force_int_axis=True,
            large_abs_safety_factor=2,
            safety_factor_scale="log",
        ),
        where=np_frontend_helpers.where(),
    )
    def test_numpy_fmin(
        dtype_x1_x2_axis,
        frontend,
        test_flags,
        fn_tree,
        on_device,
        where,
    ):
>       input_dtypes, x1, x2, axis = dtype_x1_x2_axis
E       ValueError: not enough values to unpack (expected 4, got 3)
E       Falsifying example: test_numpy_fmin(
E           dtype_x1_x2_axis=(['float16'], [array([-1.], dtype=float16)], 0),
E           where=[array([False])],
E           fn_tree='ivy.functional.frontends.numpy.fmin',
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=1,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E           ),
E           frontend='numpy',
E           on_device='cpu',
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.68.2', b'AXicY2BkAAMoBacBAEcABA==') as a decorator on your test case

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:82: ValueError
==================================================================== warnings summary =====================================================================
ivy/__init__.py:1164
  /workspaces/ivy/ivy/__init__.py:1164: UserWarning: jax module has been imported while ivy doesn't import it without setting a backend, ignore if that's intended
    warnings.warn(

ivy_tests/test_ivy/helpers/function_testing.py:69: 87 warnings
  /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:69: DeprecationWarning: `np.bool8` is a deprecated alias for `np.bool_`.  (Deprecated NumPy 1.24)
    import ivy.compiler.compiler as ic

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================= short test summary info =================================================================
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py::test_numpy_fmin[cpu-ivy.functional.backends.numpy-False-False] - ValueError: not enough values to unpack (expected 4, got 3)
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py::test_numpy_fmin[cpu-ivy.functional.backends.jax-False-False] - ValueError: not enough values to unpack (expected 4, got 3)
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py::test_numpy_fmin[cpu-ivy.functional.backends.tensorflow-False-False] - ValueError: not enough values to unpack (expected 4, got 3)
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py::test_numpy_fmin[cpu-ivy.functional.backends.torch-False-False] - ValueError: not enough values to unpack (expected 4, got 3)
============================================================= 4 failed, 88 warnings in 4.05s ==============================================================
```

**Comment by kurshakuz** [comment]
Could you create an issue for your PR, comment it to the corresponding To-Do list and link the issue to the PR as outlined in the Frontend APIs part of the [Open Tasks section](https://lets-unify.ai/ivy/contributing/open_tasks.html#open-tasks) in the docs as well?

**Comment by AryanSharma9917** [comment]
> Could you create an issue for your PR, comment it to the corresponding To-Do list and link the issue to the PR as outlined in the Frontend APIs part of the [Open Tasks section](https://lets-unify.ai/ivy/contributing/open_tasks.html#open-tasks) in the docs as well?

I created the issue and added it to to-do. Is there anything to do please guide me out.

**Comment by kurshakuz** [comment]
@AryanSharma9917 your implementation is still producing following errors' https://github.com/unifyai/ivy/pull/11829#pullrequestreview-1343073859

Please fix it and then I can review your PR

**Comment by AryanSharma9917** [comment]
> @AryanSharma9917 your implementation is still producing following errors' [#11829 (review)](https://github.com/unifyai/ivy/pull/11829#pullrequestreview-1343073859)
> 
> Please fix it and then I can review your PR

Hi sir I made the changes I think keepdims argument was giving the error

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hey! This is looking good! Can you please finally run the test locally and show the passing test for your implementation? Thanks!

**Comment by AryanSharma9917** [comment]
> Hey! This is looking good! Can you please finally run the test locally and show the passing test for your implementation? Thanks!

Hello sir, sorry for this late reply but I am facing some problems with the local setup of ivy my docker was giving some issues and when I switched to GitHub codespace I am again facing some issues can you take a look and help me out? 
image

**Comment by AryanSharma9917** [comment]
> > Hey! This is looking good! Can you please finally run the test locally and show the passing test for your implementation? Thanks!
> 
> Hello @kurshakuz  sir, sorry for this late reply but I am facing some problems with the local setup of ivy my docker was giving some issues and when I switched to GitHub codespace I am again facing some issues can you take a look and help me out? image

**Comment by AryanSharma9917** [comment]
> > > Hey! This is looking good! Can you please finally run the test locally and show the passing test for your implementation? Thanks!
> > 
> > 
> > Hello @kurshakuz  sir, sorry for this late reply but I am facing some problems with the local setup of ivy my docker was giving some issues and when I switched to GitHub codespace I am again facing some issues can you take a look and help me out? image

Hi @kurshakuz sir is there any way you can help me out, It would be very helpful.

**Comment by AryanSharma9917** [comment]
Heyy @kurshakuz could you please review my PR

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hey @AryanSharma9917! Sorry for a delay. I see that you are using Codespaces and are not able to run the tests. What I would suggest to you is to run the following pytest directly to test your implementation from the root of your Ivy folder:
```
pytest ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py::test_numpy_fmin
```

It will show you the results of your tests. Currently it is failing due to following errors:
```
E   AttributeError: module 'ivy_tests.test_ivy.test_frontends.test_numpy.helpers' has no attribute 'dtypes_values_casting_2'
```

**Comment by AryanSharma9917** [comment]
> Hey @AryanSharma9917! Sorry for a delay. I see that you are using Codespaces and are not able to run the tests. What I would suggest to you is to run the following pytest directly to test your implementation from the root of your Ivy folder:
> 
> ```
> pytest ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py::test_numpy_fmin
> ```
> 
> It will show you the results of your tests. Currently it is failing due to following errors:
> 
> ```
> E   AttributeError: module 'ivy_tests.test_ivy.test_frontends.test_numpy.helpers' has no attribute 'dtypes_values_casting_2'
> ```

Okey sir

**Comment by AryanSharma9917** [comment]
Heyy @kurshakuz I made the changes that were causing the error. Could you please review my PR again.

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py [hunk 2]
```diff
@@ -58,6 +58,46 @@ def test_numpy_minimum(
     )
 
 
+# fmin
+@handle_frontend_test(
+    fn_tree="numpy.fmin",
+    dtypes_values_casting=np_frontend_helpers.dtypes_values_casting_dtype(
+        arrs_shapes=[((2, 3),), ((2, 3),)],
```
Comment (kurshakuz) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:65]: This line is causing your test to fail:

```
========================================================================= ERRORS =========================================================================
___________________ ERROR collecting ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py ____________________
ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:64: in <module>
    dtypes_values_casting=np_frontend_helpers.dtypes_values_casting_dtype(
E   TypeError: dtypes_values_casting_dtype() got an unexpected keyword argument 'arrs_shapes'
-------------------------------------------------------------------- Captured stderr ---------------------------------------------------------------------
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
================================================================ short test summary info =================================================================
ERROR ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py - TypeError: dtypes_values_casting_dtype() got an unexpected keyword argument 'arrs_shapes'
```

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hi! Your implemenation is still failing, can you please fix and run the test locally and show me the output that you get?

For reference, I would suggest you to take a look at the `dtypes_values_casting_dtype` helper that you used: https://github.com/unifyai/ivy/blob/master/ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py#L240

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py (hunk 2 above)
Comment (AryanSharma9917) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:65]: > This line is causing your test to fail:
> 
> ```
> ========================================================================= ERRORS =========================================================================
> ___________________ ERROR collecting ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py ____________________
> ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py:64: in <module>
>     dtypes_values_casting=np_frontend_helpers.dtypes_values_casting_dtype(
> E   TypeError: dtypes_values_casting_dtype() got an unexpected keyword argument 'arrs_shapes'
> -------------------------------------------------------------------- Captured stderr ---------------------------------------------------------------------
> No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
> ================================================================ short test summary info =================================================================
> ERROR ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_extrema_finding.py - TypeError: dtypes_values_casting_dtype() got an unexpected keyword argument 'arrs_shapes'
> ```

okey sir

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**